### PR TITLE
Implement Batch polling

### PR DESCRIPTION
**Pull request by Sudakatux**
Implement batch polling

### CODE REVIEW

File: src/task/Poller.ts [hunk 1]
```diff
@@ -32,87 +33,62 @@ export class Poller {
...
-        this.concurrentCalls.push(this.singlePoll());
-      }
-    }
-  };
+    while (this.isPolling) {
+      // Concurrency could have been updated. Accounting for that
+      const count =
+        this.options.concurrency > this._tasksInProcess
+          ? this.options.concurrency - this._tasksInProcess
+          : this._tasksInProcess;
```
Comment (boney9) [inline src/task/Poller.ts:76]: const count = Math.max(0, this.options.concurrency - this._tasksInProcess);

// Math.max(0, -10) = 0
// Math.max(0, 1) = 1

File: src/task/Poller.ts [hunk 2]
```diff
@@ -32,87 +33,62 @@ export class Poller {
...
+      const taskResult: T[] = await this.pollFunction(count);
+
+      this._tasksInProcess = taskResult.length;
```
Comment (boney9) [inline src/task/Poller.ts:79]: this._tasksInProgress = this._tasksInProgress + taskResult.length;

File: src/task/Poller.ts [hunk 3]
```diff
@@ -32,87 +33,62 @@ export class Poller {
...
-      }
-    }
-  };
+    while (this.isPolling) {
+      // Concurrency could have been updated. Accounting for that
+      const count =
+        this.options.concurrency > this._tasksInProcess
+          ? this.options.concurrency - this._tasksInProcess
+          : this._tasksInProcess;
+      const taskResult: T[] = await this.pollFunction(count);
```
Comment (boney9) [inline src/task/Poller.ts:77]: Don't call if count is 0

File: src/task/TaskRunner.ts [hunk 4]
```diff
@@ -92,23 +93,15 @@ export class TaskRunner {
...
-      } else {
-        this.logger.debug(`No tasks for ${this.worker.taskDefName}`);
-      }
-    } catch (unknownError: unknown) {
-      this.handleUnknownError(unknownError);
-      this.errorHandler(unknownError as Error);
-    }
+  private batchPoll = async (count: number): Promise<Task[]> => {
+    const { workerID } = this.options;
+    const task = await this.taskResource.batchPoll(
```
Comment (boney9) [inline src/task/TaskRunner.ts:98]: tasks

File: src/task/Poller.ts [hunk 5]
```diff
@@ -5,25 +5,26 @@ interface PollerOptions {
...
-  }> = [];
-  private pollFunction: () => Promise<void> = async () => {};
+export class Poller<T> {
+  private timeoutHandler?: NodeJS.Timeout;
+  private pollFunction: (count: number) => Promise<T[]>;
+  private performWorkFunction: (work: T) => Promise<void> = async () => {};
   private polling = false;
+  private _tasksInProcess = 0;
   options: PollerOptions = {
     pollInterval: 1000,
```
Comment (v1r3n) [inline src/task/Poller.ts:15]: This might be too low --- other SDKs make this at default to 100.  Is this in millis?

Comment (Sudakatux) [reply src/task/Poller.ts:18]: yes its millis

**Review by boney9 — APPROVED** [review]

**Pull request merged**